aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -172,15 +187,23 @@ public void write(byte[] b, int off, int len) throws IOException {
 
   private void newStream() throws IOException {
     if (stream != null) {
-      stream.close();
+      closeStream();
     }
 
     createStagingDirectoryIfNotExists();
     currentStagingFile = File.createTempFile("s3fileio-", ".tmp", stagingDirectory);
     currentStagingFile.deleteOnExit();
-    stagingFiles.add(currentStagingFile);
 
-    stream = new CountingOutputStream(new BufferedOutputStream(new FileOutputStream(currentStagingFile)));
+    stagingFiles.add(new FileAndDigest(currentStagingFile));
+
+    if (isEtagCheckEnabled) {
+      currentPartMessageDigest.reset();
```

Comment:
This overlaps with Ryan's comment below, but if we just create a message digest and associate it here, we don't need to update it in the close method below.

Something like:

```java

stagingFiles.add(new FileAndDigest(currentStagingFile, newMessageDigest));
```

It means that we have to create a new digest object per file but also means they're standalone and don't require as much coordination.

Reply:
This is what I had before, but changed that to instead keep bytes as per @rdblue comment.

Reply:
@danielcweeks @rdblue either we can create a new MessageDigest for each file and use MessageDigest in `FileAndDigest` or we can use bytes. Former is likely more readable and later is likely more memory efficient. I feel memory efficiency here is not the biggest concern, so I am happy to go back to what I had originally. However, let me know what you folks think.

Reply:
I don't think we need to worry too much about the extra bytes necessary to store the digest.  Readability/maintainability is much more important than that little bit of efficiency.